### ISSUE TITLE
Make the TsMorphMetadataProvider - ProjectOptions configurable

### ISSUE THREAD

**Issue by ml1nk** [mikro-orm/mikro-orm#3168]
**Is your feature request related to a problem? Please describe.**
If ts-morph can't resolve IdentifiedReference and it is used in union with null then tsType.getText(property) returns 'any'. This case happens if mikro-orm is used indirectly with esm/node16.

**Describe the solution you'd like**
A way to configure the ts-morph ProjectOptions from the outside. It would like to add it myself but it is unclear to me how you would integrate additional options.

**Additional context**
In my casePackage B depends on Package A depends on mikro-orm. Package A and B are ECMAScript modules and tsconfig module is set to node16. Package A reexport mikro-orm with all it's interfaces to prevent version mismatch. B defines an entity like:
~~~ts
@Entity()
export class TestEntity {

  @PrimaryKey()
    id!: string

  @ManyToOne()
    parent!: IdentifiedReference<TestEntity> | null
}
~~~
which contains an nullable IdentifiedReference. On start it crashes with 'MetadataError:
~~~
Entity 'any' was not discovered, please make sure to provide it in 'entities' array when initializing the ORM (used in TestEntity.parent)' 
~~~
It is fixed after the following change in TsMorphMetadataProvider:
~~~ts
        this.project = new ts_morph_1.Project({
            compilerOptions: {
                strictNullChecks: true,
                module: ModuleKind.Node16 
            }
        });
~~~
You can see the differences if you enable logging (this.project.enableLogging(true)).

IdentifiedReference can be resolved

~~~ts
IdentifiedReference<TestEntity>
--> getType / getText
({ id: string; } & import("<path to mikro/core>").Reference<TestEntity>)
~~~

~~~ts
IdentifiedReference<TestEntity> | null
--> getType / getText
({ id: string; } & import("<path to mikro/core>").Reference<TestEntity>) | null
~~~
___

IdentifiedReference cannot be resolved

~~~ts
IdentifiedReference<TestEntity>
--> getType / getText
IdentifiedReference<TestEntity>

IdentifiedReference<TestEntity> | null
--> getType / getText
any
~~~

___

I hope my description is good/clear enough to understand why it is sometimes necessary to configure ProjectOptions directly. I can try to create a minimal reproduction for this case, if needed.

**Comment by B4nan** [comment]
I guess we should respect the options from your tsconfig ideally? Sure we can make the options configurable, but that sounds like a half bakes solution, users would still fall into this and had to debug it to understand what's happening.

**Comment by ml1nk** [comment]
> I guess we should respect the options from your tsconfig ideally? Sure we can make the options configurable, but that sounds like a half baked solution, users would still fall into this and had to debug it to understand what's happening.

You are right, if we could use the module option from tsconfig then it would most likely work out of the box in most if not all cases.

By the way, I think the problem is, that only if node16 is used the exports field is respected without needing a correspondent typesVersion. In my example Package A hasn't a typesVersion, because it is not needed anymore (and somewhat weaker than what is possible with exports).

**Comment by B4nan** [comment]
Looks like this is backwards compatible, tried to use `module: ModuleKind.Node16 ` in a project that uses TS 4.1 and it was working fine, so will do it like that to keep things simple. If there will be any issues with that, we can refactor.
